### ISSUE TITLE
Improve behaviour for Kotlin / kapt projects

### ISSUE THREAD

**Issue by chadlwilson** [dependency-check/dependency-check-gradle#239]
### Summary

When running OWASP dependency check on a Gradle Kotlin project using `kapt` annotation processing, it seems to create a **lot** of false positives in the report. These seems to stem from the `_classStructurekaptKotlin` and `_classStructurekaptTestKotlin` dependency configurations added by the plugin.

The false positives seem to arise because the plugin has lost the link from a given dependency jar to the actual Maven package it came from. For example the only suppression option you get given on the report is via `SHA1`, rather than via `packageUrl`, as you would "normally" get for a Gradle dependency or its transitives.

### Actual results

Here's the simplest re-prod I can come up with so far (happens to be a Micronaut application)

**build.gradle.kts**
```kts
plugins {
    id("org.jetbrains.kotlin.jvm") version "1.6.0"
    id("org.jetbrains.kotlin.kapt") version "1.6.0"
    id("org.jetbrains.kotlin.plugin.allopen") version "1.6.0"
    id("io.micronaut.application") version "3.0.1"
    id("org.owasp.dependencycheck") version "6.5.0.1"
}

repositories {
    mavenCentral()
}

micronaut {
    version("3.2.1")
    runtime("netty")
}

dependencies {
    kapt("io.micronaut:micronaut-http-validation")
}
```

Gives result
```
One or more dependencies were identified with known vulnerabilities in owasp-dep-kapt:

micronaut-buffer-netty-3.2.1.jar (cpe:2.3:a:netty:netty:3.2.1:*:*:*:*:*:*:*) : CVE-2014-3488, CVE-2015-2156, CVE-2019-16869, CVE-2019-20444, CVE-2019-20445, CVE-2021-21290, CVE-2021-21295, CVE-2021-21409, CVE-2021-37136, CVE-2021-37137
micronaut-http-netty-3.2.1.jar (cpe:2.3:a:netty:netty:3.2.1:*:*:*:*:*:*:*) : CVE-2014-3488, CVE-2015-2156, CVE-2019-16869, CVE-2019-20444, CVE-2019-20445, CVE-2021-21290, CVE-2021-21295, CVE-2021-21409, CVE-2021-37136, CVE-2021-37137
micronaut-http-server-netty-3.2.1.jar (cpe:2.3:a:netty:netty:3.2.1:*:*:*:*:*:*:*) : CVE-2014-3488, CVE-2015-2156, CVE-2019-16869, CVE-2019-20444, CVE-2019-20445, CVE-2021-21290, CVE-2021-21295, CVE-2021-21409, CVE-2021-37136, CVE-2021-37137
```

These are false positives because `micronaut-*-netty` is different to the core Netty Project.

![image](https://user-images.githubusercontent.com/29788154/145683819-5ed806fa-b823-4f19-8a0d-5d1f80260c02.png)

![image](https://user-images.githubusercontent.com/29788154/145683847-ea03a02b-c4ad-43a8-85d9-0705bc813467.png)

### Expected Results

Normally with regular configurations OWASP dependency check does not create these false positives, and they go away if I suppress just these two kapt configurations noted below.

I don't really understand why Dependency Check is losing the link to the actual package here. I assume this is why it is just doing a simple file name match and getting the wrong CPE.

### Workaround

One possible workaround (below) is to skip these configurations entirely, but it might be nicer if there was better support out-of-the-box somehow - if possible? It's also possible that either kapt or micronaut plugins are causing some kind of hijinx here.

```kotlin
dependencyCheck {
    skipConfigurations = listOf("_classStructurekaptKotlin", "_classStructurekaptTestKotlin")
}
```

**Comment by jeremylong** [comment]
Actually - I think this is "working as designed".  We likely need to improve the list of configurations that are just always skipped. The configurations you've listed above - and I believe you mentioned in another issue that PMD and Checkstyle also create their own scope.

**Comment by chadlwilson** [comment]
It's different to the other issue with additional configurations, because in those cases the PMD and checkstyle dependencies are real dependencies, and the vulnerabilities reported against them are typically real, not false positives. it's up to the user to decide if they care about them/want them reported or not.

In this case, including these kapt configurations cause it to report _completely incorrect vulnerabilities_ that are not real, against _all_ configurations. `micronaut-buffer-netty` for example does not (incorrectly) match to `cpe:2.3:a:netty:netty:3.2.1:*:*:*:*:*:*:*` when `micronaut-buffer-netty` is part of a regular `compileClasspath` or `runtimeClasspath` configurations.

Note that it reports the false positive against ALL configurations. If one excludes the kapt configurations it stops reporting against `compileClasspath`, `runtimeClasspath` etc. Surely that cannot be intentional/designed, that skipping configuration A could affect the vulnerabilities reported against configuration B? That's what is happening here.

I think there is something special/different about these Kapt configurations that is somehow confusing OWASP dependency check. Perhaps the answer is to always skip them, however the false positives against unrelated configurations seems to imply a deeper problem somewhere?

**Comment by cljvf9** [comment]
+1, however the only new info I can bring is that micronaut can probably be crossed off the list of possible culprits since the project where I experienced this uses spring boot.  When enabling kapt it causes the dependency check to trigger on several cve's that don't even seem to be relevant with respect to the dependencies that it flagged.  @chadlwilson 's workaround makes these false positives disappear in my case as well.  Given that, it definitely feels like a weird interaction between the kapt and dependency check plugins, but not sure what causes it or where a fix could go.

**Comment by jeremylong** [comment]
So - have you ever gone back and looked at code and wondered what you were thinking when you wrote it?  This bug has nothing to with `micronaught` or `kapt`. I'm working on a fix for this and it will be included in the 7.0.0 release.

**Comment by chadlwilson** [comment]
> So - have you ever gone back and looked at code and wondered what you were thinking when you wrote it? This bug has nothing to with `micronaught` or `kapt`. I'm working on a fix for this and it will be included in the 7.0.0 release.

Many times 😅 thanks @jeremylong !

At least the workaround is pretty simple and low risk/side effects from my perspective.
